### PR TITLE
Upgrade embedded Kotlin to 1.9.0

### PR DESCRIPTION
**Pull request by eskatos**
* See https://github.com/JetBrains/kotlin/releases/tag/v1.9.0
* Fix https://github.com/gradle/gradle/issues/25517
* Fix https://github.com/gradle/gradle/issues/24816
* Required for https://github.com/gradle/gradle/issues/23488

Kotlin language/api levels kept at 1.8 for backwards compatibility.

This PR updates the wrapper to a branch snapshot and should be followed by a regular wrapper update once available.

### CODE REVIEW

**Comment by asodja** [comment]
Can we also remove this line now:
https://github.com/gradle/gradle/blob/07cd5f40fb5feb9341614db2c2734e4aa0689431/build-logic/jvm/src/main/kotlin/gradlebuild.unittest-and-compile.gradle.kts#L227-L231

**Comment by eskatos** [comment]
@bot-gradle test RFN

**Comment by bric3** [comment]
@eskatos You may want to follow this ticket on Kotlin side as well 
* [KT-58987: Use some available JVM target if there's no JvmTarget for the inferred toolchain version ](https://youtrack.jetbrains.com/issue/KT-58987)

Currently the same issue will happen with Kotlin 1.9.0 when running on JDK 21 (or any version higher than 20).

**Comment by eskatos** [comment]
@bot-gradle test RFN

**Comment by eskatos** [comment]
@bric3, thanks, yep I'm aware of this issue. We have a similar strategy implemented for a while now for the jvm target used when compiling build scripts.

**Comment by eskatos** [comment]
@bot-gradle test RFN

**Comment by eskatos** [comment]
@bot-gradle test ACTR

**Comment by eskatos** [comment]
@bot-gradle test APT

**Comment by eskatos** [comment]
Re-enabling Kotlin compilation avoidance for projects using the `kotlin-dsl` plugin brought back [performance](https://builds.gradle.org/buildConfiguration/Gradle_Master_Check_PerformanceTestSlowLinux_Trigger/68281913?buildTab=report_project104067_Performance) of the `buildSrc abi change | largeJavaMultiProjectKotlinDsl | BuildSrcApiChangePerformanceTest` scenario. It is ~10% faster than before this PR.

**Comment by eskatos** [comment]
I'm merging this optimistically and will do the required wrapper dance. Please review, I'll handle feedback in a subsequent PR if any.

@bot-gradle test and merge

**Comment by eskatos** [comment]
@bot-gradle test and merge

**Comment by eskatos** [comment]
@bot-gradle cancel

**Review by jbartok — APPROVED** [review]
LGTM

**Review by lkasso — APPROVED** [review]

File: subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/support/KotlinCompiler.kt [hunk 1]
```diff
@@ -391,7 +391,7 @@ fun JavaVersion.toKotlinJvmTarget(): JvmTarget {
     // JvmTarget.fromString(JavaVersion.majorVersion) works from Java 9 to Java 19
```
Comment (octylFractal) [inline subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/support/KotlinCompiler.kt:391]: Comment needs updating.

Comment (eskatos) [reply subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/support/KotlinCompiler.kt]: Good catch, thanks!

**Comment by eskatos** [comment]
@bot-gradle test and merge

**Pull request merged**